src/RingGrids/general.jl

```diff
@@ -369,32 +369,29 @@ Base.BroadcastStyle(::Type{Grid}) where {Grid<:AbstractGridArray{T, N, ArrayType
 
 # allocation for broadcasting, create a new Grid with undef of type/number format T
 function Base.similar(bc::Broadcasted{AbstractGridArrayStyle{N, Grid}}, ::Type{T}) where {N, Grid, T}
-    # this escapes for Array and JLArray
-    # if isstructurepreserving(bc) || fzeropreserving(bc) 
-    #     return Grid(Array{T}(undef, size(bc)...))
-    # end
-    # return similar(convert(Broadcasted{DefaultArrayStyle{ndims(bc)}}, bc), T)
-
-    # this doesn't escape for Array but for JLArray
     return Grid(Array{T}(undef, size(bc)...))
 end
 
 # ::Val{0} for broadcasting with 0-dimensional, ::Val{1} for broadcasting with vectors, etc
 AbstractGridArrayStyle{N, Grid}(::Val{M}) where {N, Grid, M} = AbstractGridArrayStyle{N, Grid}()
 
 ## GPU
-function Broadcast.BroadcastStyle(
-    ::Type{Grid},
-) where {Grid <: AbstractGridArray{T, N, ArrayType}} where {T, N, ArrayType <: GPUArrays.AbstractGPUArray}
-    return Broadcast.BroadcastStyle(ArrayType)
-end
+struct AbstractGPUGridArrayStyle{N, ArrayType, Grid} <: GPUArrays.AbstractGPUArrayStyle{N} end
 
-function GPUArrays.backend(
-    ::Type{Grid},
-) where {Grid <: AbstractGridArray{T, N, ArrayType}} where {T, N, ArrayType <: GPUArrays.AbstractGPUArray}
+Base.BroadcastStyle(::Type{Grid}) where {Grid<:AbstractGridArray{T, N, ArrayType}} where {T, N, ArrayType <: GPUArrays.AbstractGPUArray} =
+    AbstractGPUGridArrayStyle{N, ArrayType, nonparametric_type(Grid)}()
+
+# ::Val{0} for broadcasting with 0-dimensional, ::Val{1} for broadcasting with vectors, etc
+AbstractGPUGridArrayStyle{N, ArrayType, Grid}(::Val{M}) where {N, ArrayType, Grid, M} = AbstractGPUGridArrayStyle{N, ArrayType, Grid}()
+
+function GPUArrays.backend(::Type{Grid}) where {Grid <: AbstractGridArray{T, N, ArrayType}} where {T, N, ArrayType <: GPUArrays.AbstractGPUArray}
     return GPUArrays.backend(ArrayType)
 end
 
+function Base.similar(bc::Broadcasted{AbstractGPUGridArrayStyle{N, ArrayType, Grid}}, ::Type{T}) where {N, ArrayType, Grid, T}
+    return Grid(T.(ArrayType(undef, size(bc)...)))
```

Comment:
The `T.(ArrayType(undef, size(bc)...))` is not ideal of course. It's the same problem again with determining the non-parametric type. We can keep it if we don't care about the minor performance, or we define the `nonparametric_type` function for `CuArray` and `JLArray` as well (the latter only in the tests, to not have a dependency).

Reply:
Ah yes because `ArrayType{T}` doesn't work because `ArrayType` is already parametric? I see the problem.

Reply:
Yep

Reply:
`nonparametric_type` is really something that Julia Base should have automatically for every parametric type.

Reply:
Added with [fd6d9d8](https://github.com/SpeedyWeather/SpeedyWeather.jl/pull/520/commits/fd6d9d8fdf7065577be505ee729d1b3a0dc24093)